### ISSUE TITLE
Implement ProtocolProperties validation mechanism

### ISSUE THREAD

**Issue by cloudxxx8** [edgexfoundry/device-sdk-go#649]
When a Device is created in the core-metadata, it will callback to the Device Service for awareness.
We should provide a mechanism in Device Service to validation the ProtocolProperties in the Device just created, because core-metadata doesn't have the knowledge to validate the ProtocolProperties well.

**Comment by hahattan** [comment]
what validation do we need to perform on the ProtocolProperties ?

**Comment by cloudxxx8** [comment]
We don't the validation rule, so we should provide a mechanism to call the actual Device Service.
It could be a new func in ProtocolDriver

**Comment by iain-anderson** [comment]
Considered at WG meeting 26th April:
- Define optional interface for PP validation (just as discovery is optional)
- SDK will query this on incoming requests on the callback interface where devices are added or updated
- This will allow the SDK to indicate to core-metadata that the proposed addition/update should be rejected
- Need to define an appropriate error response object (http return code would be 400?)
- Defer to post-2.0 (should be implementable without breaking change)

**Comment by jpwhitemn** [comment]
If implemented correctly - this should not be a breaking/non-compatible change from v2 (per Kamakura planning meeting).

**Comment by hahattan** [comment]
> * This will allow the SDK to indicate to core-metadata that the proposed addition/update should be rejected
> * Need to define an appropriate error response object (http return code would be 400?)

Metadata adds/updates device and calls back to device service asynchronously, should we also update the workflow there ?
For example trigger callback first and then add device into Metadata ?

**Comment by cloudxxx8** [comment]
It should be different API calls from my opinion, because we need to validate protocol property first and notify (callback) Device Service after in case the database persistence failed.
Also, this protocol property validation is not a mandatory for all the Device Services.

**Comment by hahattan** [comment]
Or making callback synchronously, and remove the device from DB if validation failed? (need a dedicated 4xx response to indicate this)

@lenny-intel @iain-anderson any thoughts ?

**Comment by lenny-intel** [comment]
> It should be different API calls from my opinion, because we need to validate protocol property first and notify (callback) Device Service after in case the database persistence failed.
> Also, this protocol property validation is not a mandatory for all the Device Services.

@hahattan , I agree with @cloudxxx8 . This is how we previously discussed handling the validation. To avoid breaking changes the SDK need to make the implementation of the `ProtocolValidation` interface optional. Similar to how we have implementing the `ProtocolDiscovery` interface as optional.

**Comment by hahattan** [comment]
@lenny-intel  I just recalled that discovery API returns 501 if not implemented:
https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/device-sdk/2.1.0#/default/post_discovery

Should we follow the same design ?

**Comment by lenny-intel** [comment]
@hahattan , IMO not in this case. With this the not implemented is a valid option.
